lib/linked_list.rb

```diff
@@ -12,120 +12,404 @@ def initialize(value, next_node = nil)
+  # returns the data value and not the node
+  def find_max
+    current = @head
+    if current == nil
+      return nil
+    elsif current.next == nil
+      return current.data
     end
-
-    # method to return the min value in the linked list
-    # returns the data value and not the node
-    def find_min
-      raise NotImplementedError
+    
+    temp = current.data
+    
+    until current.next == nil do
+      if current.data > temp
+        temp = current.data
+      end
+      current = current.next
     end
-
-
-    # method that returns the length of the singly linked list
-    def length
-      raise NotImplementedError
+    
+    if current.data > temp
+      return current.data
+    else
+      return temp
     end
-
-    # method that returns the value at a given index in the linked list
-    # index count starts at 0
-    # returns nil if there are fewer nodes in the linked list than the index value
-    def get_at_index(index)
-      raise NotImplementedError
+    
+  end
+  
+  # method to return the min value in the linked list
+  # returns the data value and not the node
+  def find_min
+    current = @head
+    if current == nil
+      return nil
+    elsif current.next == nil
+      return current.data
     end
-
-    # method to print all the values in the linked list
-    def visit
-      raise NotImplementedError
+    
+    temp = current.data
+    
+    until current.next == nil do
+      if current.data < temp
+        temp = current.data
+        current = current.next
+      else
+        current = current.next
+      end
     end
-
-    # method to delete the first node found with specified value
-    def delete(value)
-      raise NotImplementedError
+    
+    if current.data < temp
+      return current.data
+    else
+      return temp
     end
-
-    # method to reverse the singly linked list
-    # note: the nodes should be moved and not just the values in the nodes
-    def reverse
-      raise NotImplementedError
+  end
+  
+  
+  # method that returns the length of the singly linked list
+  def length
+    current = @head
+    if current == nil
+      return 0
+    elsif current.next == nil
+      return 1
     end
-
-
-    ## Advanced Exercises
-    # returns the value at the middle element in the singly linked list
-    def find_middle_value
-      raise NotImplementedError
+    
+    counter = 1
+    
+    until current.next == nil do 
+      current = current.next
+      counter += 1
     end
-
-    # find the nth node from the end and return its value
-    # assume indexing starts at 0 while counting to n
-    def find_nth_from_end(n)
-      raise NotImplementedError
+    
+    return counter
+    
+  end
+  
+  # method that returns the value at a given index in the linked list
+  # index count starts at 0
+  # returns nil if there are fewer nodes in the linked list than the index value
+  def get_at_index(index)
+    current = @head
+    if index == 0

```

Comment:
What if the list is empty?

```suggestion
    return nil if current.nil?
    if index == 0
```